internal/worker/apiremotecaller/remote.go

```diff
@@ -0,0 +1,240 @@
+	// to the remote API server. If the connection is lost or the remote server
+	// is unreachable, the channel will be closed.
+	Connection() (api.Connection, bool)
+
+	// Tag returns the tag of the remote API server.
+	Tag() names.Tag
+}
+
+// RemoteServer represents the public interface of the worker
+// responsible modeling the remote API server.
+type RemoteServer interface {
+	worker.Worker
+	RemoteConnection
+	UpdateAddresses(addresses []string)
+}
+
+// RemoteServerConfig defines all the attributes that are needed for a
+// RemoteServer.
+type RemoteServerConfig struct {
+	Clock  clock.Clock
+	Logger logger.Logger
+
+	Target names.Tag
+
+	// APIInfo is initially populated with the addresses of the target machine.
+	APIInfo *api.Info
+}
+
+// remoteServer is a worker that provides addresses for the target API server.
+type remoteServer struct {
+	tomb tomb.Tomb
+
+	target names.Tag
+	info   *api.Info
+
+	logger logger.Logger
+	clock  clock.Clock
+
+	changes chan []string
+
+	mu                sync.Mutex
+	currentConnection api.Connection
+}
+
+// NewRemoteServer creates a new RemoteServer that will connect to the remote
+// apiserver and pass on messages to the pubsub endpoint of that apiserver.
+func NewRemoteServer(config RemoteServerConfig) RemoteServer {
+	w := &remoteServer{
+		target:  config.Target,
+		info:    config.APIInfo,
+		logger:  config.Logger,
+		clock:   config.Clock,
+		changes: make(chan []string),
+	}
+	w.tomb.Go(w.loop)
+	return w
+}
+
+// Connection returns a channel that will be populated with the connection to
+// the remote API server. If the connection is lost or the remote server is
+// unreachable, the channel will be closed.
+func (w *remoteServer) Connection() (api.Connection, bool) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	return w.currentConnection, w.currentConnection != nil
+}
+
+// Tag returns the tag of the remote API server.
+func (w *remoteServer) Tag() names.Tag {
+	return w.target
+}
+
+// UpdateAddresses will update the addresses held for the target API server.
+func (w *remoteServer) UpdateAddresses(addresses []string) {
+	select {
+	case <-w.tomb.Dying():
+		return
+	case w.changes <- addresses:
+	}
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *remoteServer) Kill() {
+	w.tomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *remoteServer) Wait() error {
+	return w.tomb.Wait()
+}
+
+func (w *remoteServer) loop() error {
+	defer w.closeCurrentConnection()
+
+	ctx, cancel := w.scopedContext()
+	defer cancel()
+
+	var connected bool
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+
+		case addresses := <-w.changes:
+			w.logger.Debugf("addresses for %q have changed: %v", w.target, addresses)
+
+			// If the addresses already exist, we don't need to do anything.
+			if connected && w.addressesAlreadyExist(addresses) {
+				continue
+			}
+
+			if err := w.connect(ctx, addresses); err != nil {
```

Comment:
Connect can block indefinitely. Context passed in here should cancel if new addresses are sent on `w.changes`.

Reply:
Good point, yes, the might be wise.